lib/Dialect/RTL/Ops.cpp

```diff
@@ -255,6 +288,108 @@ static LogicalResult verifyRTLInstanceOp(RTLInstanceOp op) {
+    if (attrDict.getNamed(resultNameKey))
+      continue;
+
+    StringAttr resultName = StringAttr::get(p.getResultName(i).first, ctxt);
+    attrDict.push_back(NamedAttribute(resultNameKey, resultName));
+  }
+  return success();
+}
+
+void printResultNames(OpAsmPrinter &p, RTLInstanceOp *op) {
+  SmallVector<StringRef, 8> elideFields = {"instanceName", "moduleName"};
+
+  NamedAttrList attrs = op->getAttrs();
+  for (auto it = attrs.begin(), e = attrs.end(); it < e; ++it) {
+    StringRef attrName = it->first.strref();
+    if (attrName.startswith("rtl.rname")) {
+      auto numberStr = attrName.substr(9);
+      size_t resultIdx;
+      if (numberStr.getAsInteger<size_t>(10, resultIdx))
+        continue;
+
+      SmallString<32> resultNameStr;
+      llvm::raw_svector_ostream tmpStream(resultNameStr);
+      p.printOperand(op->getResult(resultIdx), tmpStream);
+      auto expectedName = it->second.dyn_cast<StringAttr>();
+      if (expectedName &&
+          tmpStream.str().drop_front() == expectedName.getValue()) {
+        elideFields.push_back(attrName);
+      }
+    }
+  }
+  p.printOptionalAttrDict(op->getAttrs(), elideFields);
+}
+
+void RTLInstanceOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
+  NamedAttrList attrs = getAttrs();
+  for (auto it = attrs.begin(), e = attrs.end(); it < e; ++it) {
+    if (it->first.strref().startswith("rtl.rname")) {
+      auto numberStr = it->first.strref().substr(9);
+      size_t resultIdx;
+      if (!numberStr.getAsInteger<size_t>(10, resultIdx))
+        setNameFn(getResult(resultIdx),
+                  it->second.dyn_cast<StringAttr>().getValue());
+    }
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// RTLOutputOp
+//===----------------------------------------------------------------------===/
+
+/// Verify that the num of operands and types fit the declared results.
+static LogicalResult verifyOutputOp(OutputOp *op) {
+  OperandRange outputValues = op->getOperands();
+  Operation *moduleOp = op->getParentWithTrait<OpTrait::FunctionLike>();
+
+  // Check that our region has results
+  if (!moduleOp) {
+    op->emitOpError("operation expected to be in FunctionLike region.");
+    return failure();
+  }
+
+  // Check that the we (rtl.output) have the same number of operands as our
+  // region has results.
+  FunctionType modType = getModuleType(moduleOp);
+  ArrayRef<Type> modResults = modType.getResults();
+  if (modResults.size() != outputValues.size()) {
+    op->emitOpError("must have same number of operands as region results.");
+    return failure();
+  }
+
+  // Check that the types of our operands and the region's results match.
+  for (size_t i = 0, e = modResults.size(); i < e; ++i) {
+    if (modResults[i] != outputValues[i].getType()) {
+      op->emitOpError(llvm::formatv("output types must match module. In "
+                                    "operand {0}, expected {1} but got {2}.",
```

Comment:
Instead of llvm::format, plz use something like:

```
op->emitOpError("output types must match module. In operand ") << i << ", expected " << ...
```

Reply:
Didn't even know this was possible! The extent to which MLIR/LLVM leverage operator overloading constantly amazes me.